In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# щоб графіки відображались прямо в ноутбуці
%matplotlib inline

Завантаження датасету

In [ ]:
# Завантаження основного датасету Stack Overflow Developer Survey
import pandas as pd

df = pd.read_csv("survey_results_public.csv")

# Перевіряємо перші 5 рядків датасету
df.head()

/tmp/ipython-input-1958549013.py:4: DtypeWarning: Columns (56,74,92,97,98,105,109,110,132,162,165) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("survey_results_public.csv")


ResponseId                      MainBranch              Age  \
0           1  I am a developer by profession  25-34 years old   
1           2  I am a developer by profession  25-34 years old   
2           3  I am a developer by profession  35-44 years old   
3           4  I am a developer by profession  35-44 years old   
4           5  I am a developer by profession  35-44 years old   

                                           EdLevel  \
0  Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   
1              Associate degree (A.A., A.S., etc.)   
2     Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
3     Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
4  Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   

                                          Employment  \
0                                           Employed   
1                                           Employed   
2  Independent contractor, freelancer, or self-em...   
3                                           Employed   
4  Independent contractor, freelancer, or self-em...   

                                    EmploymentAddl  WorkExp  \
0  Caring for dependents (children, elderly, etc.)      8.0   
1                                              NaN      2.0   
2                                None of the above     10.0   
3                                None of the above      4.0   
4  Caring for dependents (children, elderly, etc.)     21.0   

                                     LearnCodeChoose  \
0  Yes, I am not new to coding but am learning ne...   
1  Yes, I am not new to coding but am learning ne...   
2  Yes, I am not new to coding but am learning ne...   
3  Yes, I am not new to coding but am learning ne...   
4  No, I am not new to coding and did not learn n...   

                                           LearnCode  \
0  Online Courses or Certification (includes all ...   
1  Online Courses or Certification (includes all ...   
2  Online Courses or Certification (includes all ...   
3  Other online resources (e.g. standard search, ...   
4                                                NaN   

                                         LearnCodeAI  ...  \
0  Yes, I learned how to use AI-enabled tools for...  ...   
1  Yes, I learned how to use AI-enabled tools for...  ...   
2  Yes, I learned how to use AI-enabled tools for...  ...   
3  Yes, I learned how to use AI-enabled tools for...  ...   
4  Yes, I learned how to use AI-enabled tools for...  ...   

  AIAgentOrchestration  AIAgentOrchWrite AIAgentObserveSecure AIAgentObsWrite  \
0            Vertex AI               NaN                  NaN             NaN   
1                  NaN               NaN                  NaN             NaN   
2                  NaN               NaN                  NaN             NaN   
3                  NaN               NaN                  NaN             NaN   
4                  NaN               NaN                  NaN             NaN   

                                    AIAgentExternal AIAgentExtWrite  \
0                                           ChatGPT             NaN   
1                                               NaN             NaN   
2  ChatGPT;Claude Code;GitHub Copilot;Google Gemini             NaN   
3                               ChatGPT;Claude Code             NaN   
4                                               NaN             NaN   

                                             AIHuman  \
0                    When I don’t trust AI’s answers   
1  When I don’t trust AI’s answers;When I want to...   
2  When I don’t trust AI’s answers;When I want to...   
3  When I don’t trust AI’s answers;When I want to...   
4                    When I don’t trust AI’s answers   

                                              AIOpen  ConvertedCompYearly  \
0              Troubleshooting, profiling, debugging              61256.0   
1                          All skills. AI is a flop.             104413.0   
2  Understand how things actually work, pr

Основний датасет Stack Overflow Developer Survey успішно завантажено за допомогою бібліотеки Pandas.
Дані містять відповіді респондентів на широкий спектр питань, включаючи професійний досвід, освіту, технології та використання інструментів.

In [ ]:
# Завантажуємо файл зі структурою датасету (опис колонок)
schema = pd.read_csv("survey_results_schema.csv")

# Переглядаємо перші рядки schema
schema.head(10)

qid           qname                                           question  \
0  QID18   TechEndorse_1  What attracts you to a technology or causes yo...   
1  QID18   TechEndorse_2  What attracts you to a technology or causes yo...   
2  QID18   TechEndorse_3  What attracts you to a technology or causes yo...   
3  QID18   TechEndorse_4  What attracts you to a technology or causes yo...   
4  QID18   TechEndorse_5  What attracts you to a technology or causes yo...   
5  QID18   TechEndorse_6  What attracts you to a technology or causes yo...   
6  QID18   TechEndorse_7  What attracts you to a technology or causes yo...   
7  QID18   TechEndorse_8  What attracts you to a technology or causes yo...   
8  QID18   TechEndorse_9  What attracts you to a technology or causes yo...   
9  QID18  TechEndorse_13  What attracts you to a technology or causes yo...   

  type                                      sub  sq_id  
0   RO  AI integration or AI Agent capabilities    1.0  
1   RO                          Easy-to-use API    2.0  
2   RO                  Robust and complete API    3.0  
3   RO     Customizable and manageable codebase    4.0  
4   RO                   Reputation for quality    5.0  
5   RO      Connected to an open-source project    6.0  
6   RO              Good brand and public image    7.0  
7   RO              Reliability and low latency    8.0  
8   RO                     Costs are manageable    9.0  
9   RO                  Other (please specify):   10.0

In [ ]:
# Розмір schema: кількість питань та їх описів
schema.shape

(139, 6)

# **Завдання 1. Підрахунок загальної кількості респондентів**

In [ ]:
# Загальна кількість респондентів

total_respondents = df.shape[0]
print(total_respondents)

49191


**Висновок:** Загальна кількість респондентів, які взяли участь в опитуванні Stack Overflow Developer Survey, становить 49191 осіб, що свідчить про великий обсяг вибірки та дозволяє проводити подальший статистичний аналіз даних.

# **Завдання 2. Аналіз повноти відповідей респондентів**

In [ ]:
# Отримуємо список усіх питань з schema
questions = schema['qname'].unique()

print(f"Загальна кількість питань у schema: {len(questions)}")
questions[:10]

Загальна кількість питань у schema: 139


array(['TechEndorse_1', 'TechEndorse_2', 'TechEndorse_3', 'TechEndorse_4',
       'TechEndorse_5', 'TechEndorse_6', 'TechEndorse_7', 'TechEndorse_8',
       'TechEndorse_9', 'TechEndorse_13'], dtype=object)

In [ ]:
# Інтерсекція множин: питання + колонки датасету

schema_questions_set = set(questions)
dataset_columns_set = set(df.columns)

common_questions = schema_questions_set.intersection(dataset_columns_set)

print(f"Кількість питань, які є і в schema, і в датасеті: {len(common_questions)}")

Кількість питань, які є і в schema, і в датасеті: 126


In [ ]:
# Відбираємо тільки питання з інтерсекції
df_questions = df[list(common_questions)]

df_questions.shape

(49191, 126)

In [ ]:
# Фільтрація респондентів без пропущених значень
complete_respondents = df_questions.dropna()

complete_respondents.shape

(0, 126)

In [ ]:
# Кількість респондентів, які відповіли на всі питання
num_complete_respondents = complete_respondents.shape[0]

print(num_complete_respondents)

0


Додатково було проаналізовано кількість респондентів,
які заповнили не менше 90% питань опитування.

In [ ]:
# Частка заповнених відповідей для кожного респондента
fill_ratio = df_questions.notna().mean(axis=1)

# Респонденти з >= 90% заповнених відповідей
respondents_90 = df_questions[fill_ratio >= 0.9]

respondents_90.shape[0]

11

**Висновок:** У результаті аналізу повноти відповідей було встановлено, що жоден респондент не надав відповіді на всі запитання опитування без пропущених значень.
Це пояснюється великою кількістю умовних та необов’язкових питань у Stack Overflow Developer Survey, а також різною релевантністю окремих блоків питань для різних груп респондентів.
Отриманий результат є типовим для масштабних опитувань і свідчить про реалістичність та складність структури даних.

# **Завдання 3. Статистичний аналіз досвіду респондентів**

In [ ]:
# Аналіз досвіду роботи респондентів (WorkExp)

# Перевіряємо тип даних та кількість пропущених значень
df['WorkExp'].describe()

count    42893.000000
mean        13.367403
std         10.800117
min          1.000000
25%          5.000000
50%         10.000000
75%         20.000000
max        100.000000
Name: WorkExp, dtype: float64

In [ ]:
# Видаляємо пропущені значення для коректного статистичного аналізу
work_exp = df['WorkExp'].dropna()

# Обчислення мір центральної тенденції
mean_exp = work_exp.mean()
median_exp = work_exp.median()
mode_exp = work_exp.mode()[0]  # беремо перше значення моди

# Виводимо результати
mean_exp, median_exp, mode_exp

(np.float64(13.367402606485907), 10.0, np.float64(10.0))

In [ ]:
stats_workexp = pd.DataFrame({
    'Показник': ['Середнє (Mean)', 'Медіана (Median)', 'Мода (Mode)'],
    'Значення': [mean_exp, median_exp, mode_exp]
})

stats_workexp

Показник   Значення
0    Середнє (Mean)  13.367403
1  Медіана (Median)  10.000000
2       Мода (Mode)  10.000000

**Висновок:**
Аналіз досвіду роботи респондентів показав, що типовий учасник опитування має близько 10 років професійного досвіду. Середнє значення є дещо вищим за медіану, що свідчить про наявність меншої кількості респондентів із дуже великим стажем роботи.

# **Завдання 4. Аналіз віддаленої роботи**

In [ ]:
# Перевіряємо можливі формати роботи
df['RemoteWork'].value_counts(dropna=False)

RemoteWork
NaN                                                                             15411
Remote                                                                          10931
Hybrid (some remote, leans heavy to in-person)                                   6732
In-person                                                                        6042
Hybrid (some in-person, leans heavy to flexibility)                              5831
Your choice (very flexible, you can come in when you want or just as needed)     4244
Name: count, dtype: int64

In [ ]:
remote_mask = df['RemoteWork'].str.contains(
    'Remote',
    case=False,
    na=False
)

remote_count = remote_mask.sum()

print(f"Кількість респондентів, які працюють віддалено: {remote_count}")


Кількість респондентів, які працюють віддалено: 17663


**Висновок:** За результатами аналізу встановлено, що 17663 респондент працює у форматі повністю віддаленої роботи (Remote). Це свідчить про значну поширеність дистанційного формату зайнятості серед учасників опитування Stack Overflow Developer Survey.

# **Завдання 5. Визначення популярності Python**

In [ ]:
# Визначаємо колонку з мовами програмування
'LanguageHaveWorkedWith'
df['LanguageHaveWorkedWith'].head()

0                     Bash/Shell (all shells);Dart;SQL
1                                                 Java
2                  Dart;HTML/CSS;JavaScript;TypeScript
3                                      Java;Kotlin;SQL
4    C;C#;C++;Delphi;HTML/CSS;Java;JavaScript;Lua;P...
Name: LanguageHaveWorkedWith, dtype: object

In [ ]:
# Загальна кількість респондентів
total_respondents = df.shape[0]

# Визначаємо, хто використовує Python
python_users = df['LanguageHaveWorkedWith'].str.contains(
    'Python',
    case=False,
    na=False
)

# Кількість респондентів, які працюють з Python
num_python_users = python_users.sum()

# Обчислюємо відсоток
python_percentage = (num_python_users / total_respondents) * 100

# Виводимо результат
print(f"Кількість респондентів, які програмують на Python: {num_python_users}")
print(f"Відсоток респондентів, які програмують на Python: {python_percentage:.2f}%")

Кількість респондентів, які програмують на Python: 18466
Відсоток респондентів, які програмують на Python: 37.54%


**Висновок:** За результатами аналізу даних Stack Overflow Developer Survey встановлено, що 37,54% респондентів використовують мову програмування Python. Це свідчить про високу популярність Python серед розробників та його широке застосування в сучасній ІТ-індустрії. Отримані результати підтверджують важливу роль Python у таких напрямах, як аналіз даних, машинне навчання та автоматизація.

# **Завдання 6. Аналіз шляхів навчання програмуванню**

In [ ]:
# Перевіряємо значення у колонці LearnCode
df['LearnCode'].head()

0    Online Courses or Certification (includes all ...
1    Online Courses or Certification (includes all ...
2    Online Courses or Certification (includes all ...
3    Other online resources (e.g. standard search, ...
4                                                  NaN
Name: LearnCode, dtype: object

In [ ]:
# Фільтруємо респондентів, які навчались через онлайн курси
online_courses_df = df[
    df['LearnCode'].str.contains(
        'Online course',
        case=False,
        na=False
    )
]

num_online_course_users = online_courses_df.shape[0]

print(f"Кількість респондентів, які навчались через онлайн-курси: {num_online_course_users}")

Кількість респондентів, які навчались через онлайн-курси: 21212


**Висновок:**
Аналіз показав, що 21212 респонденти навчалися програмуванню за допомогою онлайн курсів, що свідчить про високу популярність дистанційних освітніх платформ серед розробників та зростаючу роль онлайн-освіти у сфері ІТ.

# **Завдання 7. Географічний аналіз компенсації Python-розробників**

In [ ]:
# Фільтруємо респондентів, які програмують на Python
python_devs = df[
    df['LanguageHaveWorkedWith']
    .str.contains('Python', case=False, na=False)
]
#Обираємо потрібні колонки
python_devs_comp = python_devs[['Country', 'ConvertedCompYearly']]
#Прибираємо пропущені значення компенсації
python_devs_comp = python_devs_comp.dropna(subset=['ConvertedCompYearly'])
#Групування та агрегація
country_compensation = (
    python_devs_comp
    .groupby('Country')['ConvertedCompYearly']
    .agg(
        mean_compensation='mean',
        median_compensation='median'
    )
    .reset_index()
)
#Сортування для наочності
country_compensation = country_compensation.sort_values(
    by='median_compensation',
    ascending=False
)
country_compensation = country_compensation.reset_index(drop=True)
country_compensation.head(10)


Country  mean_compensation  median_compensation
0                      Oman      390135.000000             390135.0
1                   Andorra      226103.500000             226103.5
2  United States of America      173298.590211             150000.0
3                    Israel      135828.365385             142594.0
4               Switzerland      156456.600000             142592.0
5                   Nomadic      120131.571429             139218.0
6                   Ireland      120523.918919             116015.0
7                Luxembourg      116014.714286             109054.0
8                Kyrgyzstan      106008.500000             106008.5
9                    Belize      102121.000000             102121.0

**Висновок:** Аналіз показує, що медіанна компенсація Python-розробників суттєво відрізняється між країнами. Найвищі значення спостерігаються у країнах з розвиненими IT-ринками та високою вартістю життя, що свідчить про сильну залежність рівня доходу від географічного фактору.

# **Завдання 8. Аналіз освіти найбільш оплачуваних спеціалістів**

In [ ]:
top_cols = ['ResponseId', 'EdLevel', 'ConvertedCompYearly']

# прибираємо рядки без компенсації
df_comp = df[top_cols].dropna(subset=['ConvertedCompYearly'])

# сортуємо за компенсацією у спадному порядку та беремо топ-5
top5 = df_comp.sort_values(
    by='ConvertedCompYearly',
    ascending=False
).head(5)

# скидаємо індекс для чистого вигляду таблиці
top5.reset_index(drop=True)


ResponseId                                          EdLevel  \
0       34268              Associate degree (A.A., A.S., etc.)   
1       28701  Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   
2       43144              Associate degree (A.A., A.S., etc.)   
3       35354     Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
4       45972  Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   

   ConvertedCompYearly  
0           50000000.0  
1           33552715.0  
2           18387548.0  
3           15430267.0  
4           13921760.0

In [ ]:
top5['ConvertedCompYearly'].is_monotonic_decreasing

True

**Висновок**:  У результаті аналізу встановлено, що серед 5 респондентів з найбільшою річною компенсацією переважають спеціалісти з рівнем освіти Master’s degree та Associate degree, також представлений рівень Bachelor’s degree. Це свідчить про те, що вищі рівні освіти частіше зустрічаються серед найбільш високооплачуваних респондентів опитування Stack Overflow Developer Survey, хоча наявність високих доходів можлива й без обов’язково максимального академічного ступеня.

# **Завдання 9. Аналіз популярності Python по віковим категоріям**

In [ ]:
# 1. Фільтр: хто використовує Python
df['UsesPython'] = df['LanguageHaveWorkedWith'].str.contains(
    'Python',
    case=False,
    na=False
)

# 2. Групування по вікових категоріях
age_python_stats = (
    df
    .groupby('Age')
    .agg(
        total_respondents=('ResponseId', 'count'),
        python_users=('UsesPython', 'sum')
    )
    .reset_index()
)

# 3. Розрахунок відсотка Python-розробників
age_python_stats['python_percentage'] = (
    age_python_stats['python_users']
    / age_python_stats['total_respondents']
) * 100

# 4. Залишаємо тільки потрібні колонки
age_python_stats = age_python_stats[
    ['Age', 'python_percentage']
].sort_values(by='python_percentage', ascending=False)

age_python_stats


Age  python_percentage
0    18-24 years old          40.000000
3    45-54 years old          38.629482
4    55-64 years old          37.242955
1    25-34 years old          36.939282
2    35-44 years old          36.719281
5  65 years or older          31.634820
6  Prefer not to say          31.216931

**Висновок:** У результаті аналізу популярності Python за віковими категоріями було встановлено, що найвищий відсоток Python-розробників спостерігається у віковій групі 18–24 роки. Це свідчить про високу популярність Python серед молодих спеціалістів та початківців у програмуванні. Водночас у старших вікових групах Python також залишається затребуваною мовою, що підтверджує її універсальність і широке застосування.

# **Завдання 10. Аналіз індустрій серед високооплачуваних віддалених працівників**

In [ ]:
# 1. Видаляємо пропущені значення компенсації
df_comp = df.dropna(subset=['ConvertedCompYearly'])

# 2. Обчислюємо 75-й перцентиль компенсації
comp_75 = df_comp['ConvertedCompYearly'].quantile(0.75)

print(f"75-й перцентиль річної компенсації: {comp_75:.2f}")

# 3. Фільтруємо високооплачуваних віддалених працівників
high_paid_remote = df_comp[
    (df_comp['ConvertedCompYearly'] >= comp_75) &
    (df_comp['RemoteWork'] == 'Remote')
]

# 4. Аналіз популярності індустрій
industry_stats = (
    high_paid_remote['Industry']
    .value_counts()
    .reset_index()
)

industry_stats.columns = ['Industry', 'Count']

# 5. Виводимо топ-10 індустрій
industry_stats.head(10)


75-й перцентиль річної компенсації: 120596.00


Industry  Count
0                        Software Development   1186
1                                     Fintech    190
2                                  Healthcare    188
3                                      Other:    176
4  Internet, Telecomm or Information Services    138
5                  Banking/Financial Services     88
6                                  Government     78
7                Media & Advertising Services     75
8                Retail and Consumer Services     65
9             Transportation, or Supply Chain     63

**Висновок:** У результаті аналізу було встановлено, що серед високооплачуваних віддалених працівників домінують індустрії Software Development, Fintech та Healthcare. Також суттєву частку становлять Internet, Telecom or Information Services і Banking/Financial Services. Це свідчить про те, що саме технологічні та фінансові сектори найчастіше поєднують високий рівень компенсації з віддаленим форматом роботи.

# **Загальний висновок:**
 У межах проєкту було проведено комплексний аналіз результатів опитування Stack Overflow Developer Survey, що охоплює 49 191 респондента та відображає сучасний стан ІТ-індустрії. Аналіз структури даних показав високу різноманітність запитань і наявність пропущених значень, що є типовим для масштабних міжнародних опитувань. Дослідження професійного досвіду респондентів засвідчило, що середній стаж роботи становить близько 13 років, при медіанному значенні 10 років. Також встановлено високу поширеність віддаленого формату роботи, який обрали понад 10 тисяч респондентів, що підтверджує тенденцію до зростання гнучких моделей зайнятості в ІТ-сфері. *

Окрему увагу було приділено аналізу використання мови програмування Python. Результати показали, що Python застосовують 37,54% опитаних, що підтверджує його провідну роль у сучасній розробці, аналізі даних та суміжних галузях. Географічний аналіз компенсацій Python-розробників виявив значні міжкраїнні відмінності рівня доходів, тоді як аналіз освіти найбільш високооплачуваних спеціалістів показав переважання ступенів рівня Bachelor’s та Master’s. Загалом отримані результати демонструють взаємозв’язок між досвідом, освітою, технологіями та рівнем компенсації, а також відображають ключові тенденції розвитку глобального ринку праці в ІТ-індустрії.